Support For Vanity Urls

SrS2225a opened the pull request:
Allows you (the user) to rename the random ID that uploads are assigned and give them a special vanity/custom url! As example: If  XBackBone generated a link such as https://media.nyxgoddess.org/talI2/DIsexobi55.mp4, but want something a little more custom, you can easily change that link to https://media.nyxgoddess.org/talI2/amongus.mp4. 

Additionally, if you try to change the generated link to a invalid url such as  "https://media.nyxgoddess.org/talI2/among us.mp4"  XBackBone will automatically add a `-` to make it valid. Like so: https://media.nyxgoddess.org/talI2/among-us.mp4

SrS2225a commented:
Oh no, I see. Should I worry about these failed tests and clean up those code quality issues before you merge it? Or will the current changes be sufficient enough?

SergiX44 commented:
> Oh no, I see. Should I worry about these failed tests and clean up those code quality issues before you merge it? Or will the current changes be sufficient enough?

Yes please try to make the CI pass, at least the stuff related to code style 🙏

SrS2225a commented:
Hi, I have fixed those styling issues and the CI was able to pass as normal. Do you need anything else?

SergiX44 commented:
> Hi, I have fixed those styling issues and the CI was able to pass as normal. Do you need anything else?

Thanks! I'll review the PR asap

SrS2225a commented:
Hi,

It's been a while since I heard anything from this PR so I wanted to see if you were able to review it and everything was alright. Please let me know if you have any questions or want anything changed before you accept it

SergiX44 commented on app/Controllers/MediaController.php:
```diff
@@ -188,6 +188,50 @@ public function download(Request $request, Response $response, string $userCode,
+        }
+
+        $this->database->query(
+            'UPDATE `uploads` SET `code` = ? WHERE `id` = ?',
```
<!-- Issues represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 issue:** here is missing handling of collisions, so what happen if you try to set the same vanity url to multiple uploads?‏

SrS2225a replied:
I doubled checked, I already implemented a check for collisions, including if the user tried to create an empty vanity url. Here is where the check for this is:

https://github.com/SergiX44/XBackBone/pull/507/files#diff-1f64a59ac49e401b085fe29e9b0d34bb824ebd81e54c8416c2e0e2df674b14a3R221-R223

SrS2225a replied:
Ah, that was my bad. I was doing a check on the same media object for collisions and not on all of then. Fixed :+1:

SergiX44 commented on app/Controllers/MediaController.php:
```diff
@@ -188,6 +188,50 @@ public function download(Request $request, Response $response, string $userCode,
+            )->fetch();
+        }
+
+        $data = $request->getParsedBody();
+        $vanity = $data['vanity'];
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common processes. Try to leave a link to the process described so that the reader knows how to resolve the chore.  -->
**💣 chore:** ‏There is the `param` helper to accoplish this https://github.com/SergiX44/XBackBone/blob/master/app/helpers.php#L279

SrS2225a replied:
Thanks for pointing this out. I actually did not know about this function from when I looked over your code, but it works well. Fixed :+1:

SergiX44 commented on app/Controllers/MediaController.php:
```diff
@@ -188,6 +188,50 @@ public function download(Request $request, Response $response, string $userCode,
+            )->fetch();
+        }
+
+        $data = $request->getParsedBody();
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏There is no request validation here

SrS2225a replied:
Im a little confused at what you mean there is no request validation. If you mean checking if the media does exist, I already have that in my code:

https://github.com/SergiX44/XBackBone/pull/507/files#diff-1f64a59ac49e401b085fe29e9b0d34bb824ebd81e54c8416c2e0e2df674b14a3R217-R219

SergiX44 commented on app/Controllers/MediaController.php:
```diff
@@ -188,6 +188,50 @@ public function download(Request $request, Response $response, string $userCode,
+
+        $this->logger->info('User '.$this->session->get('username').' created a vanity link for media '.$media->id);
+        
+        return $response;
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**🎯 suggestion:** ‏ ‏what about returning the updated media object here? So the frontend can update the HTML accordingly

SrS2225a replied:
You mean to like this `return $media;`? Thats because it is not a valid Psr\Http\Message\ResponseInterface type which slim expects to be returned. Otherwise, it will give an 500 error

SrS2225a replied:
Nevermind, I figured out how to return the media object. Turns out you need to add `$response->getBody()->write(json_encode($media));`, for it to return data. Thought unfortunately, it does not work with an onject so I had to convert it first.

SergiX44 commented on src/js/app.js:
```diff
@@ -57,6 +59,18 @@ var app = {
+            var $callerButton = $(this);
+            $.post(window.AppConfig.base_url + '/upload/' + id + '/vanity', {vanity: vanity}, function () {
+                $callerButton.tooltip('dispose');
+                window.location.href = window.AppConfig.base_url + '/home';
```
<!-- Issues represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 issue:** ‏the issue here is if you are updating a media deep in the pagination or in the public page, you are going to be redirected to the home page and than go back where you were before. I would update the HTML using the data returned by the post request instead of performing a redirect.

SrS2225a replied:
Sorry, must of missed that before I made the PR. I had the redirect thing as a test to make sure that everything was working properly, and was going to do the frontend stuff once it was. Again, sorry for missing that detail. Fixed :+1:

SergiX44 commented on resources/templates/comp/modal_vanity.twig:
```diff
@@ -0,0 +1,19 @@
+                </button>
+            </div>
+            <div class="modal-body">
+                <input type="text" class="form-control" id="modalVanity-input" >
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏I would limit this to 64 chars also on the frontend, since is currently the hard limit

SrS2225a replied:
Interesting, I did not know this was a hard limit with XBackbone. I thought it would be cool if users could make a vanity url as long as they like, but it makes since why this is disallowed. Fixed :+1:

SergiX44 reviewed (changes requested):
Nice start! There are some things that should be considered tho

SrS2225a commented:
Thanks, glad you like the PR! Sorry about the issues tho. I will start working on solving all of the requested fixes. I will add a comment when it is complete, or when I have any other further comments starting out from the easiest one to fix. Once all of them are fixed I will add a separate comment letting you know

SrS2225a commented:
Hi,

I have fixed all of the issues as you requested and pushed the changes to my fork. We should all be good to go. Please let me know if I missed anything. I also added comments in reply to your feedback on my thoughts.

SrS2225a commented:
Hello,

I wanted to see how the review was going as I have not heard anything from you for this merge request in a while. Is everything alright? I want to make sure everything was good with you and have not missed anything so this feature becomes official!

xbdmHQ commented:
@SergiX44 Please respond, I know you got real life stuff you got going on.

SergiX44 commented:
I'll check this out soon, sorry for the late reply, the notification was lost in the ddos that I receive every day

SergiX44 commented on src/js/app.js:
```diff
@@ -57,6 +59,31 @@ var app = {
+                $callerButton.tooltip('dispose');
+                $('#modalVanity').modal('hide');
+                $('#modalVanity-input').val('');
+                var data = JSON.parse(data);
```
<!-- Issues represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 issue:** ‏data is redeclared, please use npm to install the dependencies and run npx grunt to execute the js compilation, it will highlight missing semicolons, redeclarations, etc.

SergiX44 commented on resources/templates/dashboard/grid.twig:
```diff
@@ -28,6 +28,7 @@
                                                     {% else %}
                                                         <a class="btn btn-sm btn-info publish-toggle" data-toggle="tooltip" title="{{ lang('publish') }}" data-id="{{ media.id }}" data-published="{{ media.published }}"><i class="fas fa-check-circle"></i></a>
                                                     {% endif %}
+                                                    <button class="btn btn-info btn-sm public-vanity" data-link="{{ route('upload.delete', {'id': media.id}) }}" data-id="{{ media.id }}" data-toggle="tooltip" title="{{ lang('vanity') }}"><i class="fas fa-star"></i></button>
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common processes. Try to leave a link to the process described so that the reader knows how to resolve the chore.  -->
**💣 chore:** ‏The lang key `vanity` does not exists

SergiX44 commented on resources/templates/dashboard/grid.twig:
```diff
@@ -28,6 +28,7 @@
                                                     {% else %}
                                                         <a class="btn btn-sm btn-info publish-toggle" data-toggle="tooltip" title="{{ lang('publish') }}" data-id="{{ media.id }}" data-published="{{ media.published }}"><i class="fas fa-check-circle"></i></a>
                                                     {% endif %}
+                                                    <button class="btn btn-info btn-sm public-vanity" data-link="{{ route('upload.delete', {'id': media.id}) }}" data-id="{{ media.id }}" data-toggle="tooltip" title="{{ lang('vanity') }}"><i class="fas fa-star"></i></button>
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common processes. Try to leave a link to the process described so that the reader knows how to resolve the chore.  -->
**💣 chore:** ‏Please remove the wrong data-link, probably wrong copy-paste

SrS2225a replied:
Hello, I have resolved the lang key issue and corrected the grunt conflicts

DinnerBeef commented:
@SrS2225a any idea when you will get a chance to look at this? Been hoping this gets merged in for a while.

SrS2225a commented:
Ill be taking a look now. Been too busy with personal matters over the last month, so I have not been table to take a look for a while
